### ISSUE TITLE
Speech generation with default voice

### ISSUE THREAD

**Issue by bchinnari** [k2-fsa/ZipVoice#67]
Hi

1. Do we always have to give prompt wav, text, can't we generate with model's default voice
2. Doesn't it increase the latency (to some extent) if we are always providing prompt wav,text at runtime also even after fine-tuning the model for a speaker?
3. Is there a way to store speaker model  and do away with prompt wav,text at runtime ? so that we will have lower latency

**Comment by zhu-han** [comment]
1. This model has no default voice. You can modify the code to generate speech without a prompt. However, I found that speech generated without a prompt is often unsatisfactory. 
2. Yes, this will increase latency. A workaround is using a short prompt wav. If you plan to fine-tune the model for a specific speaker, you can train it without a prompt—this way, no prompt will be needed during inference either.

3. Unfortunately, our model doesn’t support this feature at present. But I’m currently working on a version supporting it, and I hope to release it soon.

**Comment by wincing2** [comment]
> 1. This model has no default voice. You can modify the code to generate speech without a prompt. However, I found that speech generated without a prompt is often unsatisfactory.
> 2. Yes, this will increase latency. A workaround is using a short prompt wav. If you plan to fine-tune the model for a specific speaker, you can train it without a prompt—this way, no prompt will be needed during inference either.
> 3. Unfortunately, our model doesn’t support this feature at present. But I’m currently working on a version supporting it, and I hope to release it soon.

2. How to train it without a prompt? 
3. is the version you're working on the same as Point 2, the train without prompt? If not, could you give a little detail about how it's done.

**Comment by zhu-han** [comment]
> > 1. This model has no default voice. You can modify the code to generate speech without a prompt. However, I found that speech generated without a prompt is often unsatisfactory.
> > 2. Yes, this will increase latency. A workaround is using a short prompt wav. If you plan to fine-tune the model for a specific speaker, you can train it without a prompt—this way, no prompt will be needed during inference either.
> > 3. Unfortunately, our model doesn’t support this feature at present. But I’m currently working on a version supporting it, and I hope to release it soon.
> 
> 2. How to train it without a prompt?
> 3. Is the version you're working on the same as Point 2, training without prompt? If not, could you give a little detail about how it's done.

2. For training, set mask_percent to (1.0, 1.0). This ensures that speech conditions are always zero.
https://github.com/k2-fsa/ZipVoice/blob/572ba69c985f7dcd0b48f0ed539a3a6116db5f58/zipvoice/models/zipvoice.py#L360
You'll need to adjust the inference code accordingly on your own.

3. I'll release it once it's finished.

**Comment by wincing2** [comment]
> > > 3. Unfortunately, our model doesn’t support this feature at present. But I’m currently working on a version supporting it, and I hope to release it soon.  
> 3. I'll release it once it's finished.

Will there be a new pretrained checkpoint released besides model update?

**Comment by Alekksander66** [comment]
Hello @zhu-han  i already found a way to "cache" prompt wav and prompt text 

if it can help you 

`#!/usr/bin/env python3
import argparse
import os
import numpy as np
import torch
import torchaudio

from tokenizer import LibriTTSTokenizer
from feature import TorchAudioFbank, TorchAudioFbankConfig

def save_prompt_cache(cache_dir, voice_name, prompt_tokens, prompt_features, prompt_features_lens, prompt_rms):
    os.makedirs(cache_dir, exist_ok=True)
    np.save(os.path.join(cache_dir, f"{voice_name}_tokens.npy"), prompt_tokens)
    np.save(os.path.join(cache_dir, f"{voice_name}_features.npy"), prompt_features.cpu().numpy())
    np.save(os.path.join(cache_dir, f"{voice_name}_features_lens.npy"), prompt_features_lens.cpu().numpy())
    np.save(os.path.join(cache_dir, f"{voice_name}_rms.npy"), np.array([prompt_rms]))

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--prompt-wav', type=str, required=True, help='wav path')
    parser.add_argument('--prompt-text', type=str, required=True, help='Transcription')
    parser.add_argument('--voice-name', type=str, required=True, help='cache name (ex: client42)')
    parser.add_argument('--out-dir', type=str, default='prompt_cache', help='cache repertory')
    parser.add_argument('--token-file', type=str, required=True, help='token.txt LibriTTS')
    parser.add_argument('--token-type', type=str, default='char', help='token type')
    parser.add_argument('--sampling-rate', type=int, default=24000)
    parser.add_argument('--feat-dim', type=int, default=100)
    parser.add_argument('--n-fft', type=int, default=1024)
    parser.add_argument('--hop-length', type=int, default=256)
    parser.add_argument('--feat-scale', type=float, default=0.1)
    parser.add_argument('--target-rms', type=float, default=0.1)
    parser.add_argument('--cuda', action='store_true')
    args = parser.parse_args()

    device = torch.device("cuda" if args.cuda and torch.cuda.is_available() else "cpu")

    # Tokenizer LibriTTS
    tokenizer = LibriTTSTokenizer(args.token_file, token_type=args.token_type)
    prompt_tokens = tokenizer.texts_to_token_ids([args.prompt_text])

    wav, sr = torchaudio.load(args.prompt_wav)
    rms = torch.sqrt(torch.mean(torch.square(wav)))
    if rms < args.target_rms:
        wav = wav * args.target_rms / rms
    if sr != args.sampling_rate:
        wav = torchaudio.transforms.Resample(sr, args.sampling_rate)(wav)

    fbank_cfg = TorchAudioFbankConfig(
        sampling_rate=args.sampling_rate,
        n_mels=args.feat_dim,
        n_fft=args.n_fft,
        hop_length=args.hop_length,
    )
    feature_extractor = TorchAudioFbank(fbank_cfg)
    features = feature_extractor.extract(wav, sampling_rate=args.sampling_rate).to(device)
    features = features.unsqueeze(0) * args.feat_scale
    features_lens = torch.tensor([features.size(1)], device=device)

    save_prompt_cache(args.out_dir, args.voice_name, prompt_tokens, features, features_lens, rms)
    print(f"✅ Cache created for voice '{args.voice_name}' in '{args.out_dir}'.")

if __name__ == "__main__":
    main()
`

The script have to be remanied to get use to the new repo, but you have the logic here if it can help you

It simply pre-save the prompt-wav and text into .npy files, that can be re-used for inferencem so that way we jump the prompt-wav and prompt-text step

**Comment by zhu-han** [comment]
I'm afraid this isn't what other users are looking for. This method only changes the way to read prompts and won't reduce inference latency.
